crates/sui-framework/packages/sui-framework/sources/k_of_n_upgrade_policy.move

```diff
@@ -0,0 +1,522 @@
+        /// the ID of the proposal (`ProposedUpgrade` instance)
+        proposal: ID,
+        /// digest of the proposal
+        digest: vector<u8>,
+        /// ballot used for this vote
+        ballot: ID,
+        /// signer of the transfer operation
+        signer: address,
+        /// original owner of the ballot, that is the address the
+        /// ballot was first sent to
+        original_owner: address,
+        /// the receiver of the ballot
+        ballot_receiver: address,
+    }
+
+    /// Record a deleted `Ballot`.
+    struct BallotDeleted has copy, drop {
+        /// the ID of the proposal (`ProposedUpgrade` instance)
+        proposal: ID,
+        /// digest of the proposal
+        digest: vector<u8>,
+        /// ballot used for this vote
+        ballot: ID,
+        /// signer of the deleted operation
+        signer: address,
+        /// original owner of the ballot, that is the address the
+        /// ballot was first sent to
+        original_owner: address,
+    }
+
+    /// Allowed voters must in the [2, 100] range.
+    const EAllowedVotersError: u64 = 0;
+    /// Required votes must be less than allowed voters.
+    const ERequiredVotesError: u64 = 1;
+    /// The signer of the proposed upgrade is not one of the allowed voters.
+    const EInvalidProposer: u64 = 2;
+    /// The `Ballot` used to vote is not for the correct proposal (`ProposedUpgrade`)
+    const EInvalidBallot: u64 = 3;
+    /// Not enough votes to perform the upgrade.
+    const ENotEnoughVotes: u64 = 4;
+    /// An upgrade was issued already, and the operation requested failed.
+    const EAlreadyIssued: u64 = 5;
+    /// The operation required the signer to be the same as the upgrade proposer.
+    const ESignerMismatch: u64 = 6;
+    /// Proposal (`KofNUpgradeCap`) and upgrade (`ProposedUpgrade`) do not match.
+    const EInvalidProposalForUpgrade: u64 = 7;
+
+    /// Create a `KofNUpgradeCap` given an `UpgradeCap`.
+    /// The returned instance is the only and exclusive controller of upgrades. 
+    /// If the transaction sender is not in the set of allowed voters it will be
+    /// added. The sender is the publisher of the package which is always allowed
+    /// to vote.
+    /// The `k` (`required_votes`) out of `n` (length of `voters`) is set up
+    /// at construction time and it is immutable.
+    public fun new(
+        cap: UpgradeCap,
+        required_votes: u64,
+        voters: VecSet<address>,
+        ctx: &mut TxContext,
+    ): KofNUpgradeCap {
+        assert!(vec_set::size(&voters) > 1, EAllowedVotersError);
+        // add the signer to the list of voters if that was not there
+        let signer = tx_context::sender(ctx);
+        if (!vec_set::contains(&voters, &signer)) {
+            vec_set::insert(&mut voters, signer);
+        };
+        assert!(vec_set::size(&voters) <= 100, EAllowedVotersError);
+        assert!(required_votes < vec_set::size(&voters), ERequiredVotesError);
```

Comment:
Why can't this be equal? (i.e. unanimous vote)?

Reply:
as the comment above, it does not have to be that way and we could do equal, I had equal until I reviewed the PR to submit and changed it then and adjust the tests.
I do not know it felt that if you want them all a true multisig transaction is better, but I can see that being arguable both ways.

Reply:
I think wanting them all may make sense too (and I would argue for it in fact) since the K of N upgrade policy buys a couple things over a pure multi-sig (IMO):
1. IIRC multi-sig has a lower bound to the number of parties involved which is lower than 100. 
2. Multi-sig is synchronous voting while this is much more asynchronous
3. Multi-sig is not as transparent as this i.e., there is no easily determinable and indelible way of recording who did/did not vote with a multi-sig, whereas with this this is a much more transparent process.

Reply:
resolving given I changed the code to follow the suggestions (equal or n out of n is fine too)